lib/block-supports/layout.php

```diff
@@ -39,9 +39,33 @@ function gutenberg_register_layout_support( $block_type ) {
  * @return string                                CSS style.
  */
 function gutenberg_get_layout_style( $selector, $layout, $has_block_gap_support = false, $gap_value = null, $should_skip_gap_serialization = false, $fallback_gap_value = '0.5em', $block_spacing = null ) {
+	gutenberg_set_layout_style( $selector, $layout, $has_block_gap_support, $gap_value, $should_skip_gap_serialization, $fallback_gap_value, $block_spacing );
+
+	// Use a unique store to avoid conflicts with other stores and implementations.
+	$store     = WP_Style_Engine_CSS_Rules_Store_Gutenberg::get_store( 'block-supports/layout/' . md5( $selector ) );
```

Comment:
Oh, this is an interesting way to solve it. Will this effectively involve registering dozens of stores if a post contains dozens of blocks that opt-in to layout support?

It looks like @ramonjd is attempting to solve a similar problem as this PR in #42452. In that PR there's a `wp_style_engine_enqueue_block_supports_styles` function ([here](https://github.com/WordPress/gutenberg/pull/42452/files#r923963122)) which would call a method of `WP_Style_Engine` to add the declarations to the `block-supports` store. If I'm following that PR correctly, when the main `WP_Style_Engine` class is constructed, it registers the hooks for output, so the behaviour is encapsulated a bit further. I _think_ the result then would be that we'd only need the one `block-supports` store, and it gets output once? Though there's a bit of discussion here (https://github.com/WordPress/gutenberg/pull/42452/files#r922931702) that for that process to work, we might need to flag at a store level whether or not it's to be enqueued?

So many little bits to figure out 😀

Reply:
> Will this effectively involve registering dozens of stores if a post contains dozens of blocks that opt-in to layout support?

Nope! The `gutenberg_get_layout_style` is no longer used anywhere and can probably be removed. I tweaked it here to use a unique store per selector just to cover all bases in case someone uses the function in a plugin etc. So it's basically just for backwards-compatibility purposes 😉

Reply:
Ah, thanks for clarifying, I was probably reading too quickly! Good idea thinking of backwards compatibility — I wonder if in practice anyone actually uses it directly? I think I'd always thought of it as an internal / de-facto private method, but yeah, there very well could be someone attempting to use it in a plugin or theme 🤔

Reply:
>  If I'm following that PR correctly, when the main WP_Style_Engine class is constructed, it registers the hooks for output, so the behaviour is encapsulated a bit further. I think the result then would be that we'd only need the one block-supports store, and it gets output once? 

That's how I've built it 👍 

Is there a reason why we'd want multiple stores for block supports? 

I'm open to accusations of egocentric bias, but I think https://github.com/WordPress/gutenberg/pull/42452 might do want we want without the need for granular stores. I've tested it using the CSS storing steps in this PR.

> we might need to flag at a store level whether or not it's to be enqueued?

I've solved this by breaking apart the compile (into CSS strings) and parser (parsing raw block style values) functionalities, which should set us up for further abstraction.

> I wonder if in practice anyone actually uses it directly? I think I'd always thought of it as an internal / de-facto private method, but yeah, there very well could be someone attempting to use it in a plugin or theme

I reckon we should encourage using the internal method in practice (in code and eventually in documentation), and I think it'd be neater for our own purposes anyway because

1. The use of the classname is abstracted
2. The store key exists in one place (fewer places to update)
3. Abstracting logic should mean that making changes (which will be inevitable during these early phases) will give us few headaches.

If theme authors really want to hook into Gutenberg's stores, they'll be able to do it via any filters we build in later as  demonstrated in https://github.com/WordPress/gutenberg/pull/42493

Reply:
> Is there a reason why we'd want multiple stores for block supports?

Ooooh! I see that we'd have to return the layout styles for backwards compatibility. 🤦 It is a bit fiddly. I'm sure we'll come up with something.